vip-datamanagement/src/main/java/fr/insalyon/creatis/vip/datamanager/server/business/ShanoirStorageBusiness.java

```diff
@@ -0,0 +1,160 @@
+            this.errorKey = errorKey;
+        }
+    }
+
+    /* GASW regexps in 3.2.0 version
+
+       local token=`echo $URI | sed -r 's/^.*[?&]token=([^&]*)(&.*)?$/\1/i'`
+       local fileName=`echo $URI | sed -r 's#^shanoir:/(//)?([^/].*)\?.*$#\2#i'`
+       local apiUrl=`echo $URI | sed -r 's/^.*[?&]apiurl=([^&]*)(&.*)?$/\1/i'`
+       local refreshToken=`echo $URI | sed -r 's/^.*[?&]refreshToken=([^&]*)(&.*)?$/\1/i'`
+       local format=`echo $URI | sed -r 's/^.*[?&]format=([^&]*)(&.*)?$/\1/i'`
+       local datasetId=`echo $URI | sed -r 's/^.*[?&]datasetId=([^&]*)(&.*)?$/\1/i'`
+       local keycloak_client_id=`echo $URI | sed -r 's/^.*[?&]keycloak_client_id=([^&]*)(&.*)?$/\1/i'`
+       local refresh_token_url=`echo $URI | sed -r 's/^.*[?&]refresh_token_url=([^&]*)(&.*)?$/\1/i'`
+
+       So objective : generate "shanoir:/fileName?apiurl=[...]&datasetId=[...]&format=[...]&token=[...]&refreshToken=[....]&keycloak_client_id=[....]&refresh_token_url=[....]
+    */
+    public String generateUri(
+            ExternalPlatform externalPlatform, String parameterName, String parameterValue) throws BusinessException {
+
+        verifyExternalPlatform(externalPlatform);
+
+        String apiUrl = externalPlatform.getUrl();
+        String keycloakClientId = externalPlatform.getKeycloakClientId();
+        String refreshTokenUrl = externalPlatform.getRefreshTokenUrl();
+
+        String token = subString(UrlKeys.TOKEN, parameterValue);
+        String refreshToken = subString(UrlKeys.REFRESH_TOKEN, parameterValue);
+        String fileName = subString(UrlKeys.FILE_NAME, parameterValue);
+
+        if(CoreConstants.RESULTS_DIRECTORY_PARAM_NAME.equals(parameterName)){
+            String type = subString(UrlKeys.TYPE, parameterValue);
+            String md5 = subString(UrlKeys.MD5, parameterValue);
+            String uploadUrl = externalPlatform.getUploadUrl();
+
+            return buildUploadUri(fileName, uploadUrl, token, refreshToken, type, md5, keycloakClientId, refreshTokenUrl);
+        }
+        
+        String format = subString(UrlKeys.FORMAT, parameterValue);
+        String datasetId = subString(UrlKeys.DATASET_ID, parameterValue);
+
+        return buildDownloadUri(datasetId, apiUrl, token, refreshToken, format, fileName, keycloakClientId, refreshTokenUrl);
+    }
+
+    private void verifyExternalPlatform(ExternalPlatform externalPlatform)
+            throws RuntimeException, BusinessException {
+        if ( ! externalPlatform.getType().equals(Type.SHANOIR)) {
+            logger.error("Trying to generate a shanoir URI for a non shanoir storage {}",
+                    externalPlatform.getType());
+            throw new BusinessException("Cannot generate shanoir uri");
+        }
+        if (externalPlatform.getUrl() == null) {
+            logger.error("A shanoir external storage must have an URL to generate an URI");
+            throw new BusinessException("Cannot generate shanoir uri");
+        }
```

Comment:
Maybe also verify `externalPlatform.getKeycloakClientId()` and `externalPlatform.getRefreshTokenUrl()` and `externalPlatform.getUploadUrl` ?